feat(iroh-relay)!: add a QUIC server for QUIC address discovery to the iroh relay.

ramfox opened the pull request:
## Description

This PR adds a QUIC endpoint to the relay server that can do QUIC address discovery. It also contains structs/functions for properly doing the Client side interaction for this process.

Also, this adjust the `RelayNode` to include configuration on how to speak to the QUIC endpoint on the relay server.

QUIC is disabled by default and requires a `TlsConfig` to be configured in order to work.

closes #2964 

## Breaking Changes

- `iroh_base::relay_map::RelayNode` now has field `quic` that takes a `Option<iroh_base::relay_map::QuicConfig>`
- `iroh::test_utils::run_relay_server_with(stun: Option<StunConfig>)` => `iroh::test_utils::run_relay_server_with(stun: Option<StunConfig>, quic: bool)`
    - when `quic` is `true`, it will start a quic server for QUIC address discovery, that has self signed tls certs for testing.
- `iroh_relay::server::ServerConfig` has field `quic` that takes a `Option<iroh_relay::server::QuicConfig>`
- `iroh_relay::server::TlsConfig.quic_bind_addr` is a new field that takes a `SocketAddr`
- `iroh_relay::server::TlsConfig.server_config` is a new field that takes a `rustls::ServerConfig`
- field `config` has been removed from variant `iroh_relay::server::CertConfig::LetsEncrypt`
- variant `iroh_relay::server::CertConfig::LetsEncrypt` has a new field `state` that takes a `tokio_rustls_acme::AcmeState<EC, EA>`
- variant `iroh_relay::server::CertConfig::Manual` no longer has field `private_key`

## Change checklist

- [x] Self-review.
- [x] Documentation updates following the [style guide](https://rust-lang.github.io/rfcs/1574-more-api-documentation-conventions.html#appendix-a-full-conventions-text), if relevant.
- [x] Tests if relevant.
- [x] All breaking changes documented.

dignifiedquire commented on iroh-base/src/relay_map.rs:
```diff
@@ -122,6 +131,15 @@ pub struct RelayNode {
+}
+
+fn quic_port() -> u16 {
+    0
```
why not use the default port?

dignifiedquire commented on iroh-net/src/defaults.rs:
```diff
@@ -1,14 +1,19 @@
+/// for QUIC address discovery
+///
+/// The port is "QUIC" typed on a phone keypad.
+pub use iroh_base::relay_map::DEFAULT_QUIC_PORT;
```
if we reexport this, maybe as sth like `DEFAULT_RELAY_QUIC_PORT`?

dignifiedquire commented on iroh-relay/src/client.rs:
```diff
@@ -377,6 +377,30 @@ impl ClientBuilder {
 }
 
+#[cfg(any(test, feature = "test-utils", feature = "dangerous-certs"))]
+/// Creates a client config that trusts relay servers without verification
```
more specifically: trusts any server without verifying their TLS certificate

dignifiedquire commented on iroh-relay/src/main.rs:
```diff
@@ -132,6 +132,27 @@ struct Config {
+    /// Will ignore any other tls certificates passed to the relay server.
+    ///
+    /// Binds the QUIC server to "127.0.0.1".
+    enable_quic_local_cert: bool,
```
maybe we should prefix this with sth like `dangerous_`?

dignifiedquire commented on iroh-relay/src/main.rs:
```diff
@@ -132,6 +132,27 @@ struct Config {
+    ///
+    /// If `None`, then the [`DEFAULT_QUIC_PORT`] will be used. If 0,
+    /// then the socket will bind to a random unused port.
+    quic_bind_port: Option<u16>,
```
just `quic_port`?

dignifiedquire commented on iroh-relay/src/main.rs:
```diff
@@ -362,46 +391,83 @@ async fn main() -> Result<()> {
+        if let Some(ref tls) = relay_tls {
+            quic_config = Some(QuicConfig::new(
+                tls.server_config.clone(),
+                tls.https_bind_addr.ip(),
```
should we rename this config value if we change its meaning?

ramfox replied:
yea i'm into the `tls.quic_bind_addr` config option

dignifiedquire commented on iroh-relay/src/main.rs:
```diff
@@ -362,46 +391,83 @@ async fn main() -> Result<()> {
+            let (_, server_config) = iroh_relay::server::self_signed_tls_certs_and_config();
+            quic_config = Some(QuicConfig::new(
+                server_config,
+                Ipv6Addr::UNSPECIFIED.into(),
```
why ipv6 by default?

ramfox replied:
quinn specific: if you bind to `Ipv6Addr::UNSPECIFIED` it will attempt to bind to both `Ipv6Addr::UNSPECIFIED` and `Ipv4Addr::UNSPECIFIED` if you can do dual-stack sockets on that machine.

dignifiedquire replied:
I see, can you add a comment about this?

dignifiedquire commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+            let cancel_accept_loop = cancel.clone();
+
+            let task = tokio::task::spawn(async move {
+            let mut set = JoinSet::new();
```
missing indent?

dignifiedquire commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+        /// Closes the underlying QUIC endpoint and the tasks running the
+        /// QUIC connections.
+        pub fn shutdown(&self) {
+            self.cancel.cancel();
```
`async` and await the task?

dignifiedquire commented on iroh-relay/src/server.rs:
```diff
@@ -122,6 +138,33 @@ pub struct StunConfig {
 }
 
+/// Configuration for the QUIC server.
+// TODO(ramfox): limits? Limiting connections or accepted # of connections from a single IP? from a single node ID?
```
we also need to likely adjust the stream/connection limits for a server here

ramfox replied:
https://github.com/n0-computer/iroh/issues/2967

flub replied:
I took the liberty to edit that issue.

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+    }
+
+    async fn handle_connection(conn: quinn::Incoming) -> Result<()> {
+        let connection = conn.await?;
```
This will trigger all the garbage that gets sent to this socket and log them as a warning by the caller.  (Quinn itself doesn't document this very well, we have at least some explanation hidden somewhere so we can say "told you so": https://docs.rs/iroh-net/latest/iroh_net/endpoint/struct.Incoming.html#method.accept).

So perhaps we can create something like:
```rust
#[thiserror::Error]
enum ConnError {
    Incoming(quinn::ConnectionError),
    Real(#[from] anyhow::Error),
}
```
so that the caller can distinguish maybe only debug-log the incoming errors, and then map the error from the incoming.

ramfox replied:
ooo okay while implementing this, I saw that we end up debug-logging all other errors that happen in this method anyway. So I'm just debug-logging all errors in this method and not returning anything.

ramfox replied:
to circle back to this, I added back `Return<()>` to the `handle_connection` method, but now we:

- log all errors in this method at the "debug" level (no more `warn` for `Incoming` errors)
- when errors are no longer handled at the call site, we just pass the `Result` along to the join set
- at the `set.join_next()` site, we only take action if the error is a `panic`, in which case we call `panic!`, otherwise the error has already been logged & we ignore it.

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+
+    async fn handle_connection(conn: quinn::Incoming) -> Result<()> {
+        let connection = conn.await?;
+        info!("established");
```
That's too high for prod, should be debug.

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+                quinn::ServerConfig::with_crypto(Arc::new(quic_config.server_config));
+            let transport_config = Arc::get_mut(&mut server_config.transport).unwrap();
+            transport_config
+                .max_concurrent_uni_streams(0_u8.into())
```
Should we also disable `max_concurrent_bidi_streams`?  That would avoid another set of buffers that can be filled by a malicious remote.  And my understanding from the protocol is that no data is ever sent.

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+        let connection = conn.await?;
+        info!("established");
+        // wait for the client to close the connection
+        let connection_err = connection.closed().await;
```
Do we need a timeout on this?  Or do we allow this to be connected forever?  I guess forever is fine and allows for more interesting use in the future.

ramfox replied:
yea, i had this same question and went for the easiest path right now. Trying to understand how this could be abused, but since we don't accept any streams, it seems like the worst someone can do is open up a ton of connections and just hang around? But then they are also hanging around with a bunch of connection opens 😂 Anyway, if you don't have a strong opinion about this right now, i'd like to leave it as is.

flub replied:
Agreed that opening a ton of connections is the biggest malicious use right now.  And that's covered by the generic "do we need connection limits" issue.  So agree that we can leave this like this for now.

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+                return Ok(());
+            }
+            _ => {
+                warn!(
```
I'd turn down this level to debug.  It is something we as the server have no control over.  While it is interesting to get this info, it is not a problem with the server.

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+            }
+            _ => {
+                warn!(
+                    "{} - error closing connection {connection_err:?}",
+                    connection.remote_address()
```
```suggestion
                    error_code, reason, "error closing connection {connection_err:#}"
```
The address is already part of the span.

You'll need to expose reason to add it as a field.

ramfox replied:
not all connection errors have error_codes or reasons, but I see what you are saying!

flub reviewed (commented):
Don't let the extensive commenting put you off, it looks good!

The config plumbing is just painful.

flub commented on example.config.toml:
```diff
@@ -1,4 +1,4 @@
-[[derp_nodes]]
+[[relay_nodes]]
```
Looking at this afterwards maybe this change is by mistake?  I don't think I've seen the code change for it anywhere.

ramfox replied:
this is just an example config file that we've had floating in our repo and it's been wrong for quite some time now...

flub replied:
oh lol, you're fixing a broken example.  :heart:

ramfox commented on iroh-net/src/test_utils.rs:
```diff
@@ -66,10 +92,14 @@ pub async fn run_relay_server_with(
         .parse()
         .unwrap();
+    let quic = Some(iroh_base::relay_map::QuicConfig {
+        port: server.quic_addr().map_or(DEFAULT_QUIC_PORT, |s| s.port()),
```
oh yea duh ty

ramfox commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+use quinn::VarInt;
+
+/// ALPN for our quic addr discovery
+pub const ALPN_QUIC_ADDR_DISC: &[u8] = b"quic";
```
oh! i need to remove this, this is unused! We don't negotiate any alpns because QAD doesn't have anything to do with what protocol you are using and is negotiated based on the quic transport parameters. We don't even send any data over, we just handshake to establish the connection and then wait for the server to send over the observed addresses.

flub replied:
I would argue there is some kind of protocol here for qad:  namely that clients nor servers are allowed to create any streams and can only connect and disconnect and keep the connection alive using normal quic transport config things.  That's a protocol.

ramfox replied:
fair. okay, i'll add in the ALPN to the endpoints.

ramfox commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+/// ALPN for our quic addr discovery
+pub const ALPN_QUIC_ADDR_DISC: &[u8] = b"quic";
+/// Endpoint close error code
+pub const QUIC_ADDR_DISC_CLOSE_CODE: VarInt = VarInt::from_u32(0);
```
changing to 1 sounds good to me

ramfox commented on iroh-relay/src/server.rs:
```diff
@@ -122,6 +138,33 @@ pub struct StunConfig {
+}
+
+impl QuicConfig {
+    /// Create a new [`QuicConfig`] from a [`TlsConfig`].
```
ahh yes, thank you. this used to be MUCH more complicated, but as I understood what was going on better, it slimmed down to this! I'll see what removing it looks like 😂

ramfox replied:
Okay, so now, with the `quic_bind_addr` in the TlsConfig, it actually makes sense to keep this and pass in a TlsConfig so that the QuicConfig is supplied with all the correct pieces.

ramfox commented on iroh-net/src/defaults.rs:
For now, net-report just ignores the `quic` field, so this shouldn't be a problem.

ramfox commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+                                     warn!(remote_address = ?remote_addr, "error handling connection {err:?}")
+                                 }
+                                 res
+                             });
```
ty for pointing this out. it isn't fully natural to me yet, but soooo obviously better when looking through logs 😂

ramfox commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,301 @@
+        let mut transport = quinn_proto::TransportConfig::default();
+        // enable address discovery
+        transport
+            .send_observed_address_reports(true)
```
I had some logic for keeping it, but I don't remember what it was, which tells me it probably wasn't very logical in the first place 😂 

removing it.

ramfox commented on iroh-relay/src/server.rs:
```diff
@@ -153,14 +198,12 @@ pub struct Limits {
     /// Use a static TLS key and certificate chain.
     Manual {
-        /// The TLS private key.
-        private_key: rustls::pki_types::PrivateKeyDer<'static>,
```
We used to generate the server config much later in the process. Once you have the certs & server config generated, you don't need the private keys anymore. the `private_key` field became unused.

flub replied:
Ah yes, makes sense.  Thanks

ramfox commented on iroh-relay/src/main.rs:
```diff
@@ -132,6 +132,27 @@ struct Config {
+    ///
+    /// If `None`, then the [`DEFAULT_QUIC_PORT`] will be used. If 0,
+    /// then the socket will bind to a random unused port.
+    quic_bind_port: Option<u16>,
```
it's now moved into the TlsConfig!

Need to do a bit more auditing to make sure that we are still binding to the addrs we want, by default, but the config is looking cleaner!

ramfox commented on iroh-relay/src/main.rs:
```diff
@@ -362,46 +391,83 @@ async fn main() -> Result<()> {
-                https_bind_addr: tls.https_bind_addr(&cfg),
-                cert: cert_config,
+async fn maybe_load_tls(
+    cfg: &Config,
```
hmmm, okay so I want to do this, but we use the `cfg: &Config` to generate the `https_bind_port` and `quic_bind_port`.

I think the params may have to stay as-is

ramfox commented:
@flub can you take a look at the config options for the relay server again.

I removed the option to generate self-signed certs automatically and added info to the readme about how to pass those in.

However, the `--dev` flag removes all tls config in the code and I wanted to keep the "simple" dev mode available for those who don't care about QUIC addr disc. SO there are now 2 dev-like flags, `--dev` that runs the relay server locally over http with QAD disabled and `--dev-quic` that expects tls to be configured for QAD, but still runs the server over http. 

Is this the worst? I tried to keep it as simple as possible 😂  but could use another set of eyes.

ramfox commented:
actually, @flub 

I'm wondering if instead, `enabled_quic_addr_discovery` should actually default to `false` and if we should add a `--quic` flag on the CLI that explicitly enables it instead of doing this weird `--dev-quic` thing.

So if you want to run QAD locally you would run:
`cargo run --bin iroh-relay -- --config-path=PATH --dev --quic`

flub commented:
> That feels cleaner? thoughts?

Hum, it is awkward that TLS certs are basically becoming mandatory even for a local dev relay server.  But it is inevitable and we'll only grow more dependence over a QUIC transport over time.

But, the good news is that we don't have to get this right from the start.  The level of stability for `--dev` really doesn't matter that much.  And as you say, right now qad is just an optional extra rather than an integral part.

So I'm fine with both of these options.  And yes, the latter one, disable qad by default, is the simpler one so why not.

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,325 @@
+                                Some(conn) => {
+                                     debug!("accepting connection");
+                                     let remote_addr = conn.remote_address();
+                                     set.spawn(async move {
+                                         handle_connection(conn).await
+                                     }.instrument(info_span!("qad-conn", %remote_addr)));
```
I think you can write this without wrapping it in another future.  Something like:
```suggestion
                                     set.spawn(
                                         handle_connection(conn)
                                         .instrument(info_span!("qad-conn", %remote_addr))
                                     );
```

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,325 @@
+                Ok(())
+            }
+            _ => {
+                debug!("error closing connection {connection_err:#?}",);
```
"connection closed with error"?

This side is not doing the active closing I guess.

ramfox replied:
tru

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,325 @@
+    ///
+    /// Any errors that happen during this connection do not need to be handled,
+    /// and will be logged at the debug level in this function.
+    async fn handle_connection(incoming: quinn::Incoming) -> Result<()> {
```
I find it a bit odd that this returns `Result` but then nothing ever does anything with it.  This is not needed for the panic case you already handle anyway.  Would it not be clearer to return nil here?  (i.e. return nothing as nil return is implicit)

ramfox replied:
good point. this was confusing; I had a hard time figuring out where the errors were handled. Switched it. Took out the debugs here and log the error in the same place we check for panics.

flub commented on iroh-relay/src/quic.rs:
```diff
@@ -0,0 +1,325 @@
+            }
+        };
+        let mut observed_addr = res.expect("checked");
+        // if we've sent an to an ipv4 address, but
```
```suggestion
        // if we've sent to an ipv4 address, but
```

flub commented on iroh-relay/src/server.rs:
```diff
@@ -438,19 +478,50 @@ impl Server {
+        (None, None, 0) => Ok(Err(anyhow!("Nothing to supervise"))),
+        (None, None, _) => tasks.join_next().await.expect("checked"),
+        (None, Some(quic_ep), 0) => quic_ep.task_handle().await.map(anyhow::Ok),
+        (None, Some(quic_ep), _) => {
```
This match explosion here is a bit unpleasant I guess.  I had to try this a few times but I think this might work to remove the match and do just a single `select!` statement for this entire supervisor:

```rust
    let quic_enabled = quic_server.is_some();
    let mut quic_fut = match quic_server {
        Some(ref mut server) => futures_util::future::Either::Left(server.task_handle()),
        None => futures_util::future::Either::Right(futures_lite::future::pending()),
    };
    tokio::select! {
        biased;
        ret = tasks.join_next(), if tasks.len() > 0 => todo!(),
        ret = &mut quic_fut, if quic_enabled => todo!(),
    }
```

The explanation:

- You can disable branches in the select using the `if`.
- A disabled branch won't poll the future from that branch.
- But even if disabled all futures are constructed so need to be valid.
- So you need a future that exists regardless of whether a server is present.
- Because of borrowing rules you need to do the `.is_some()` stuff up front.

ramfox replied:
damn, nice.

flub reviewed (approved):
Looks great!  Don't think there's anything blocking but still a few suggestions that are hopefully useful.

ramfox commented on iroh-net/src/defaults.rs:
```diff
@@ -1,14 +1,19 @@
+/// for QUIC address discovery
+///
+/// The port is "QUIC" typed on a phone keypad.
+pub use iroh_base::relay_map::DEFAULT_QUIC_PORT as DEFAULT_RELAY_QUIC_PORT;
```
😂 dig asked for the rename, since it would no longer have the `relay_map` prefix

we can also...not have this in the defaults & just import it from iroh_base where needed.

ramfox replied:
okay actually, I just renamed them all to DEFAULT_RELAY_QUIC_PORT. This isn't like STUN that has a default port, this our personal QUIC port, so it feels like addition of `RELAY` is justified.

ramfox commented on iroh-relay/src/client.rs:
```diff
@@ -377,6 +377,24 @@ impl ClientBuilder {
+/// Creates a client config that trusts any servers without verifying their TLS certificate.
+///
+/// Should be used for testing local relay setups only.
+pub fn make_dangerous_client_config() -> rustls::ClientConfig {
```
This was `pub` for the `doctor report` command for the client part in the next PR. It's where the `dangerous_certs` feature originally came from. I'm going to keep it private for now, and see what is actually needed, when the next PR settles down.

ramfox commented on iroh-relay/Cargo.toml:
```diff
@@ -129,11 +131,12 @@ server = [
 name = "iroh-relay"
 path = "src/main.rs"
-required-features = ["server"]
+required-features = ["server", "dangerous-certs"]
```
removed

ramfox commented on iroh-relay/src/client.rs:
```diff
@@ -377,6 +377,24 @@ impl ClientBuilder {
     }
 }
 
+#[cfg(any(test, feature = "test-utils", feature = "dangerous-certs"))]
```
only used in tests now, so I've removed the other features. I'll remember to add it if we end up needing the `dangerous-certs` feature in the next PR

ramfox commented on iroh-relay/src/main.rs:
```diff
@@ -115,13 +124,20 @@ struct Config {
+    ///
+    /// When the `--dev` flag is used on the CLI, it will set `http_only` to `true`.
+    #[serde(default = "cfg_defaults::http_only")]
+    http_only: bool,
```
did both. moved it TlsConfig, called it `dangerous_http_only`, and added comments explaining.

ramfox commented on iroh-relay/src/main.rs:
```diff
@@ -132,6 +148,13 @@ struct Config {
+    ///
+    /// If no `tls` is set, this will error.
+    ///
+    /// Defaults to `true`
```
Yea, that's why I thinking QUIC should be `false` on default. If someone wants to use QUIC they need to add a config file anyway, so enabling QUIC via the config at that point is easy.

ramfox commented on iroh-relay/src/main.rs:
```diff
@@ -366,45 +418,74 @@ async fn main() -> Result<()> {
+    cfg: &Config,
+) -> Result<Option<relay::TlsConfig<std::io::Error, std::io::Error>>> {
+    let Some(ref tls) = cfg.tls else {
+        println!("no tls, returning none");
```
oops! should not have made it past manual testing...😅

ramfox commented on iroh-relay/src/main.rs:
```diff
@@ -513,6 +597,8 @@ mod tests {
     #[tokio::test]
     async fn test_rate_limit_config() -> TestResult {
         let config = "
+            enable_quic_addr_discovery = false
```
no this was good, it made me realize that we should handle enabling quic and the tls config differently.
